test/e2e/firewall.go

```diff
@@ -0,0 +1,471 @@
+	framework.Logf("Waiting up to %v for satisfying expectedHosts for %v times", timeout, countToSucceed)
+	hittedHosts := sets.NewString()
+	count := 0
+	condition := func() (bool, error) {
+		var respBody bytes.Buffer
+		reached, err := testReachableHTTPWithContentTimeout(externalIP, int(httpPort), "/hostname", "", &respBody, firewallTestTcpTimeout)
+		if err != nil || !reached {
+			return false, nil
+		}
+		hittedHost := strings.TrimSpace(respBody.String())
+		if !expectedHosts.Has(hittedHost) {
+			framework.Logf("Error hitting unexpected host: %v, reset counter: %v", hittedHost, count)
+			count = 0
+			return false, nil
+		}
+		if !hittedHosts.Has(hittedHost) {
+			hittedHosts.Insert(hittedHost)
+			framework.Logf("Missing %+v, got %+v", expectedHosts.Difference(hittedHosts), hittedHosts)
+		}
+		if hittedHosts.Equal(expectedHosts) {
+			count++
+			if count >= countToSucceed {
+				return true, nil
+			}
+		}
+		return false, nil
+	}
+
+	if err := wait.Poll(time.Second, timeout, condition); err != nil {
+		return fmt.Errorf("error waiting for expectedHosts: %v, hittedHosts: %v, count: %v, expected count: %v", expectedHosts, hittedHosts, count, countToSucceed)
+	}
+	return nil
+}
+
+func getInstanceTags(cloudConfig framework.CloudConfig, instanceName string) (*compute.Tags, error) {
+	gceCloud := cloudConfig.Provider.(*gcecloud.GCECloud)
+	res, err := gceCloud.GetComputeService().Instances.Get(cloudConfig.ProjectID, cloudConfig.Zone, instanceName).Do()
+	if err != nil {
+		return nil, err
+	}
+	return res.Tags, nil
+}
+
+func setInstanceTags(cloudConfig framework.CloudConfig, instanceName string, tags []string) ([]string, error) {
+	gceCloud := cloudConfig.Provider.(*gcecloud.GCECloud)
+	// Re-get instance everytime because we need the latest fingerprint for updating metadata
+	resTags, err := getInstanceTags(cloudConfig, instanceName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get instance tags: %v", err)
+	}
+
+	_, err = gceCloud.GetComputeService().Instances.SetTags(cloudConfig.ProjectID, cloudConfig.Zone, instanceName, &compute.Tags{Fingerprint: resTags.Fingerprint, Items: tags}).Do()
+	if err != nil {
+		return nil, fmt.Errorf("failed to set instance tags: %v", err)
+	}
+	framework.Logf("Sent request to set tags %v on instance: %v", tags, instanceName)
+	return resTags.Items, nil
+}
+
+// From cluster/gce/config-test.sh, master name is set up using below format:
+// MASTER_NAME="${INSTANCE_PREFIX}-master"
+func getInstancePrefix(masterName string) (string, error) {
+	if !strings.HasSuffix(masterName, "-master") {
+		return "", fmt.Errorf("unexpected master name format: %v", masterName)
+	}
+	return masterName[:len(masterName)-7], nil
+}
+
+// From cluster/gce/config-test.sh, cluster ip range is set up using below command:
+// CLUSTER_IP_RANGE="${CLUSTER_IP_RANGE:-10.180.0.0/14}"
+func getClusterIpRange() string {
+	return "10.180.0.0/14"
```

Comment:
I think it is dangerous to hardcode this range. But I have not figured out a good approach to retrieve it properly. Advise needed.